File: client_libs/javascript/memento.js
"use strict";

const net = require("net");
const path = require("path");
const fs = require("fs");
const { ChildProcess, spawn } = require("child_process");
const { randomUUID } = require("crypto");

/** @typedef {{id: string, content: string, time: number, user: string? | undefined, score: number? | undefined, lifetime: number? | undefined}} MemObj */
/** @typedef {{memory: MemObj, dictance: number}} QueriedMemObj */

class Memory {
    /** @type {string} */ id;
    /** @type {string} */ content;
    /** @type {number} */ time;

    /** @type {string? | undefined} */ user;
    /** @type {number? | undefined} */ score;
    /** @type {number? | undefined} */ lifetime;

    /**
     *
     * @param {MemObj? | undefined} params
     */
    constructor(params = null) {
        if (params) {
            Object.assign(this, params);
        }
    }

    /**
     * @param {Record<string,any>} input
     * @returns {Memory}
     */
    static fromRecord(input) {
        return new Memory(input);
    }

    /**
     * @returns {MemObj}
     */
    toRecord() {
        let obj = {
            id: str(self.id),
            content: str(self.content),
            time: int(self.time),
        };
        if (this.user) obj["user"] = str(self.user);
        if (this.score) obj["score"] = float(self.score);
        if (this.lifetime) obj["lifetime"] = int(self.lifetime);
        return obj;
    }

    toJson() {
        return JSON.stringify(this.toRecord());
    }
};
exports.Memory = Memory

class QueriedMemory {
    /** @type {Memory} */ memory;
    /** @type {number} */ distance;

    /**
     *
     * @param {QueriedMemObj? | undefined} params
     */
    constructor(params = null) {
        if (params) {
            Object.assign(this, params);
        }
    }

    /**
     * @param {QueriedMemObj} input
     * @returns
     */
    static fromRecord(input) {
        return new QueriedMemory(input);
    }

    toRecord() {
        return {
            memory: this.memory.toRecord(),
            distance: Number(this.distance),
        };
    }

    toJson() {
        return JSON.stringify(this.toRecord());
    }
};
exports.QueriedMemory = QueriedMemory

class QueryResult {
    /** @type {QueriedMemory[]} */ shortTerm
    /** @type {QueriedMemory[]} */ longTerm
    /** @type {Memory[]} */        users
}
exports.QueryResult = QueryResult

class WSWrapper {
    constructor(url, options = {}) {
        this.url = url;
        this.websocket = null;
        this.listeners = new Map();

        // Options with sane defaults
        this.reconnectInterval = options.reconnectInterval || 2000; // ms
        this.maxRetries = options.maxRetries || Infinity;
        this.retryCount = 0;
        this.autoReconnect = options.autoReconnect !== false;

        this._connect();
    }

    _connect() {
        try {
            this.websocket = new WebSocket(this.url);

            this.websocket.onopen = (event) => {
                this.retryCount = 0;
                this._emit("open", event);
            };

            this.websocket.onmessage = (event) => {
                this._emit("message", event.data);
            };

            this.websocket.onerror = (event) => {
                this._emit("error", event);
            };

            this.websocket.onclose = (event) => {
                this._emit("close", event);

                if (this.autoReconnect && this.retryCount < this.maxRetries) {
                    setTimeout(() => {
                        this.retryCount++;
                        this._connect();
                    }, this.reconnectInterval);
                }
            };
        } catch (err) {
            this._emit("error", err);
        }
    }

    send(data) {
        if (this.websocket && this.websocket.readyState === WebSocket.OPEN) {
            this.websocket.send(data);
        } else {
            this._emit("error", new Error("WebSocket is not open"));
        }
    }

    close(code = 1000, reason = "Normal Closure") {
        this.autoReconnect = false;
        if (this.websocket) {
            this.websocket.close(code, reason);
        }
    }

    on(event, callback) {
        if (!this.listeners.has(event)) {
            this.listeners.set(event, []);
        }
        this.listeners.get(event).push(callback);
    }

    off(event, callback) {
        if (this.listeners.has(event)) {
            this.listeners.set(
                event,
                this.listeners.get(event).filter((cb) => cb !== callback)
            );
        }
    }

    _emit(event, data) {
        if (this.listeners.has(event)) {
            for (const callback of this.listeners.get(event)) {
                try {
                    callback(data);
                } catch (err) {
                    console.error("WSWrapper listener error:", err);
                }
            }
        }
    }
}

exports.Memento = class Memento {
    /** @private @type {URL} */ _url;
    /** @private @type {WSWrapper?} */ _conn;
    /** @private @type {ChildProcess} */ _proc;
    /** @private @type {string? | undefined} */ _abs_dir;

    /** @private @type {Record<string, (value: any)=>any>} */ _resolvers = {};

    /**
     * @param {{abs_dir?: string?, host: string, port: number}?} params
     */
    constructor(params = { abs_dir: null, host: "127.0.0.1", port: 4286 }) {
        this._conn = null;
        this._url = new URL(`wss://${params.host}:${params.port.toString()}`);
        this._abs_dir = params.abs_dir;
    }

    /**
     * @param {number} port
     * @param {string} host
     * @param {number} timeoutMs
     * @returns {Promise}
     */
    isPortOpen(port, host, timeoutMs = 2_000) {
        return new Promise((resolve) => {
            const socket = new net.Socket();

            socket.setTimeout(timeoutMs); // timeout in ms

            socket.once("connect", () => {
                socket.destroy();
                resolve(true); // something is listening
            });

            socket.once("timeout", () => {
                socket.destroy();
                resolve(false);
            });

            socket.once("error", () => {
                resolve(false); // connection refused or unreachable
            });

            socket.connect(port, host);
        });
    }

    async connect() {
        let isConnected = await this.isPortOpen(
            this._url.port,
            this._url.host,
            2_000
        );
        if (!isConnected) {
            if (!this._abs_path) {
                throw new Error(
                    "Memento instance is not open, and no abs_path have been provided to start a new one."
                );
            }

            pyPath = path.join(this._abs_dir, "venv", "Scripts", "python.exe");
            mainPath = path.join(this._abs_dir, "main.py");

            if (!fs.existsSync(pyPath) || !fs.existsSync(mainPath)) {
                throw new Error(
                    "abs_dir provided does not point to a valid Memento directory."
                );
            }

            this._proc = spawn(pyPath, [mainPath], {
                cwd: this._abs_dir,
            });
            process.once("exit", () => {
                this._proc.kill(2);
            });

            isConnected = await this.isPortOpen(
                this._url.port,
                this._url.host,
                10_000
            );
            if (!isConnected) {
                throw new Error("failed to run new Memento instance.");
            }
        }

        this._conn = new WSWrapper(this._url);
        this._conn.on("open", () => {
            this._conn.on("message", this._handleMessage);
        });
    }

    async _handleMessage(data) {
        try {
            var obj = JSON.parse(data);
        } catch (e) {
            return;
        }

        if (typeof obj["type"] != "string") {
            throw new Error(
                'received malformed payload from Memento. field: "type" is not of type string'
            );
        }

        if (typeof obj["uid"] != "string") {
            throw new Error(
                'received malformed payload from Memento. field: "uid" is not of type string'
            );
        }

        let msgType = obj["type"];
        let msgId = obj["uid"];

        switch (msgType) {
            case "query": {
                let res = new QueryResult()
                let dbs = obj["from"]
                
                if ("stm" in dbs)
                    for (let x of obj["stm"])
                        res.shortTerm.push(QueriedMemory.fromRecord(x))

                if ("ltm" in dbs)
                    for (let x of obj["ltm"])
                        res.longTerm.push(QueriedMemory.fromRecord(x))
                
                if ("users" in dbs)
                    for (let x of obj["users"])
                        res.users.push(Memory.fromRecord(x))
                
                if (msgId in this._resolvers) {
                    let resolver = this._resolvers[msgId]
                    resolver(res)
                } else {
                    throw new Error("received unhandled response to query request.")
                }
            }
        }
    }

    /**
     * @param {string} queryStr
     * @param params
     * @returns {Promise<QueryResult>}
     */
    async query(queryStr, params = {
        collectionName: "default",
        user: null,
        from: ["stm", "ltm", "users"],
        n: [1, 1, 1],
        timeoutMs: 5_000,
    }) {
        let reqId = randomUUID()

        let resolver = undefined
        let promise = new Promise((resolve) => {
            resolver = resolve
        })

        this._resolvers[reqId] = resolver

        this._conn.send(JSON.stringify({
            "uid": reqId,
            "type": "query",
            "query": queryStr,
            "ai_name": collectionName,
            "user": user,
            "from": from,
            "n": n,
        }))

        let timeoutPromise = new Promise((resolve) => setTimeout(resolve, params.timeoutMs))
        let result = await Promise.race([timeoutPromise, promise])

        delete this._resolvers[reqId]

        if (!result) {
            throw new Error("timeout")
        }
        return result
    }
};

File: client_libs/python/memento.py
# pip install websockets

import asyncio
import os
import socket
import subprocess
import json
import uuid

from typing import Callable, Literal
from enum import Enum

from websockets.asyncio.client import ClientConnection, connect

from typing import Optional, Self
from pydantic import BaseModel, Field


class Memory(BaseModel):
    id: str = Field(...)
    content: str = Field(...)
    time: int = Field(...)

    user: Optional[str] = Field(default=None)
    score: Optional[float] = Field(default=None)
    lifetime: Optional[int] = Field(default=None)

    class Config:
        populate_by_name = True
    

    @staticmethod
    def from_dict(input: dict)-> Self:
        return Memory(
            id=input.get("id", uuid.uuid4()),
            content=input.get("content", ""),
            time=input.get("time", 0),
            user=input.get("user", None),
            score=input.get("score", None),
            lifetime=input.get("lifetime", None),
        )

    def to_dict(self)-> dict:
        obj = {
            "id": str(self.id),
            "content": str(self.content),
            "time": int(self.time),
        }

        if self.user is not None:
            obj["user"] = str(self.user)
        if self.score is not None:
            obj["score"] = float(self.score)
        if self.lifetime is not None:
            obj["lifetime"] = int(self.lifetime)
        return obj

    def to_json(self)-> str:
        return json.dumps(self.to_dict())


class QueriedMemory(BaseModel):
    memory: Memory = Field(...)
    distance: float = Field(...)
    
    @staticmethod
    def from_dict(input: dict)-> Self:
        mem: dict | None = input.get("memory", None),
        return QueriedMemory(
            memory=Memory(
                id=mem.get("id", uuid.uuid4()),
                content=mem.get("content", ""),
                time=mem.get("time", 0),
                user=mem.get("user", None),
                score=mem.get("score", None),
                lifetime=mem.get("lifetime", None),
            ),
            distance=input.get("distance")
        )

    def to_dict(self)-> dict:
        return {
            "memory": self.memory.to_dict(),
            "distance": float(self.distance),
        }
    
    def to_json(self)-> str:
        return json.dumps(self.to_dict())


class OpenLlmMsg(BaseModel):
    role: Literal["assistant", "user", "system"] | str = Field(...)
    content: str = Field(...)
    name: Optional[str] = Field(default=None)

    class Config:
        populate_by_name = True


class DbEnum(Enum):
    SHORT_TERM = "stm"
    LONG_TERM = "ltm"
    USERS = "users"


class QueryResult:
    short_term: list[QueriedMemory]
    long_term: list[QueriedMemory]
    users: list[Memory]


class CountResult:
    short_term: int
    long_term: int


class Memento:
    _conn: ClientConnection
    _proc: subprocess.Popen
    _uri: str

    _summary_cb: Callable[[str], None] = None
    _pending_requests: dict[str, asyncio.Future] = {}


    def __init__(self, abs_dir: str = "", host: str = "127.0.0.1", port: int = 4286, loop=None):
        self._conn = None
        self._uri = f"wss://{host}:{str(port)}"

        self._proc = None
        if not self._is_port_open(host, port, timeout=2.0):
            if abs_dir == "":
                raise Exception("Memento instance is not open, and no abs_path have been provided to start a new one.")

            py_path = os.path.join(abs_dir, "venv", "Scripts", "python.exe")
            main_path = os.path.join(abs_dir, "main.py")

            if not os.path.isfile(py_path) or not os.path.isfile(main_path):
                raise Exception("abs_dir provided does not point to a valid Memento directory.")

            self._proc = subprocess.Popen([py_path, main_path], cwd=abs_dir)

            if not self._is_port_open(host, port, timeout=10.0):
                raise Exception("failed to run new Memento instance.")

        self._loop = loop or asyncio.get_event_loop()
        self._ws_task = self.loop.create_task(self._runner())
    

    def __del__(self):
        if self._proc != None:
            self._proc.kill()
        
        if self._conn != None:
            self._conn.close()
    

    def _is_port_open(host: str, port: int, timeout: float = 2.0) -> bool:
        try:
            with socket.create_connection((host, port), timeout=timeout):
                return True
        except (socket.timeout, ConnectionRefusedError, OSError):
            return False
    
    
    async def _runner(self):
        async with connect(uri=self._uri) as ws:
            self._conn = ws
            async for message in ws:
                try:
                    obj = json.loads(message)
                except Exception as e:
                    raise Exception("received malformed json from Memento.", e)
                
                if "type" not in obj:
                    raise Exception("received malformed payload from Memento. missing field: \"type\"")

                if not isinstance(obj["type"], str):
                    raise Exception("received malformed payload from Memento. field: \"type\" is not of type str")

                if "uid" not in obj:
                    raise Exception("received malformed payload from Memento. missing field: \"uid\"")
                
                if not isinstance(obj["uid"], str):
                    raise Exception("received malformed payload from Memento. field: \"uid\" is not of type str")

                message_type = obj["type"]
                message_id = obj["uid"]

                match message_type:
                    case "query":
                        res = QueryResult()
                        dbs: list[str] = obj["from"]
                        
                        if "stm" in dbs:
                            res.short_term = [QueriedMemory.from_dict(x) for x in obj["stm"]]
                        if "ltm" in dbs:
                            res.long_term = [QueriedMemory.from_dict(x) for x in obj["ltm"]]
                        if "users" in dbs:
                            res.users = [Memory.from_dict(x) for x in obj["users"]]

                        if message_id in self._pending_requests:
                            future = self._pending_requests.pop(message_id, None)
                            if future:
                                future.set_result(res)
                        else:
                            raise Exception("received unhandled response to query request.")

                    case "summary":
                        if self._summary_cb is None:
                            continue
                        
                        summary: str = obj["summary"]
                        self._summary_cb(summary)
                    
                    case "count":
                        res = CountResult()
                        
                        if "stm" in obj:
                            res.short_term = obj["stm"]
                        if "ltm" in obj:
                            res.long_term = obj["ltm"]
                        
                        if message_id in self._pending_requests:
                            future = self._pending_requests.pop(message_id, None)
                            if future:
                                future.set_result(res)
                        else:
                            raise Exception("received unhandled response to count request.")
                        
    
    
    def set_on_summary(self, cb: Callable[[str], None]):
        self._summary_cb = cb
    

    async def query(self,
            query_str: str,
            collection_name: str = "default",
            user: str | None = None,
            _from: list[DbEnum] = [DbEnum.SHORT_TERM, DbEnum.LONG_TERM, DbEnum.USERS],
            n: list[int] = [1, 1, 1],
            timeout: float = 5.0,
        )-> QueryResult:

        req_id = str(uuid.uuid4())
        future: asyncio.Future[QueryResult] = asyncio.Future()
        self._pending_requests[req_id] = future
        
        await self._conn.send(
            message=json.dumps({
                "uid": req_id,
                "type": "query",
                "query": query_str,
                "ai_name": collection_name,
                "user": user,
                "from": [x.value for x in _from],
                "n": n,
            }),
            text=True,
        )

        try:
            async with asyncio.timeout(timeout):
                return await future
        except asyncio.TimeoutError as e:
            future.set_result(None)
            raise e


    async def store(self,
            memories: list[Memory],
            collection_name: str = "default",
            to: list[DbEnum] = [DbEnum.SHORT_TERM, DbEnum.USERS],
        )-> None:

        await self._conn.send(
            message=json.dumps({
                "uid": str(uuid.uuid4()),
                "type": "store",
                "memories": [x.to_dict() for x in memories],
                "ai_name": collection_name,
                "to": [x.value for x in to],
            }),
            text=True,
        )


    async def process(self,
            messages: list[OpenLlmMsg],
            context: list[OpenLlmMsg] = None,
            collection_name: str = "default",
        )-> None:

        if context is None:
            context = []

        await self._conn.send(
            message=json.dumps({
                "uid": str(uuid.uuid4()),
                "type": "process",
                "ai_name": collection_name,
                "messages": [x.model_dump(mode="json") for x in messages],
                "context": [x.model_dump(mode="json") for x in context],
            }),
            text=True,
        )


    async def close(self)-> None:
        await self._conn.send(
            message=json.dumps({
                "uid": str(uuid.uuid4()),
                "type": "close",
            }),
            text=True,
        )
    
    async def evict(self, collection_name: str = "default")-> None:
        await self._conn.send(
            message=json.dumps({
                "uid": str(uuid.uuid4()),
                "type": "evict",
                "ai_name": collection_name,
            }),
            text=True,
        )
    
    async def clear(self,
            collection_name: str = "default",
            target: list[DbEnum] = [DbEnum.SHORT_TERM, DbEnum.LONG_TERM, DbEnum.USERS],
        )-> None:

        await self._conn.send(
            message=json.dumps({
                "uid": str(uuid.uuid4()),
                "type": "clear",
                "ai_name": collection_name,
                "target": [x.value for x in target],
            }),
            text=True,
        )
    
    async def count(self,
            collection_name: str = "default",
            _from: list[DbEnum] = [DbEnum.SHORT_TERM, DbEnum.LONG_TERM, DbEnum.USERS],
            timeout: float = 5.0,
        )-> None:

        req_id = str(uuid.uuid4())
        future: asyncio.Future[CountResult] = asyncio.Future()
        self._pending_requests[req_id] = future
        
        await self._conn.send(
            message=json.dumps({
                "uid": req_id,
                "type": "count",
                "ai_name": collection_name,
                "from": [x.value for x in _from],
            }),
            text=True,
        )

        try:
            async with asyncio.timeout(timeout):
                return await future
        except asyncio.TimeoutError as e:
            future.set_result(None)
            raise e
